### ISSUE TITLE
Cannot deploy if source path is symlink

### ISSUE THREAD

**Issue by vytautasgermanavicius** [forcedotcom/cli#1375]
### Summary

 `force:source:deploy` does not work with symlinks: neither symbolic, nor absolute.

### Steps To Reproduce:

```
$ sfdx -v
sfdx-cli/7.134.0 linux-x64 node-v16.13.2
$ ls -la "./vendors/SObjectFabricator"
lrwxrwxrwx 1 vytautas vytautas 32 saus.  12 11:19 ./vendors/SObjectFabricator -> ../modules/SObjectFabricator/src
$ sfdx force:source:deploy  -p "./vendors/SObjectFabricator"
ERROR running force:source:deploy:  /home/░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░/vendors/SObjectFabricator: Could not infer a metadata type
$ sfdx force:source:deploy  -p ./modules/SObjectFabricator/src/
*** Deploying with SOAP API v53.0 ***
Deploy ID: 0░░░░░░░░░░░░░░░D
SOURCE PROGRESS | ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ | 0/13 Components
```

### Expected result

Deployment has to work with real directories and symlinks.

### Actual result

Deploy fails, if source path is symlink. But deployment works, if directory is provided not as symlink.

Deploy fails with versions:

- sfdx-cli/7.132.0 linux-x64 node-v16.13.1
- sfdx-cli/7.134.0 linux-x64 node-v16.13.2

### System Information
```
$ sfdx version --verbose --json
{
        "cliVersion": "sfdx-cli/7.134.0",
        "architecture": "linux-x64",
        "nodeVersion": "node-v16.13.2",
        "pluginVersions": [
                "@oclif/plugin-autocomplete 0.3.0 (core)",
                "@oclif/plugin-commands 1.3.0 (core)",
                "@oclif/plugin-help 3.3.1 (core)",
                "@oclif/plugin-not-found 1.2.6 (core)",
                "@oclif/plugin-plugins 1.10.11 (core)",
                "@oclif/plugin-update 1.5.0 (core)",
                "@oclif/plugin-warn-if-update-available 1.7.3 (core)",
                "@oclif/plugin-which 1.0.4 (core)",
                "alias 1.2.0 (core)",
                "auth 1.8.0",
                "config 1.3.14 (core)",
                "generator 1.2.1 (core)",
                "info 1.2.0 (core)",
                "salesforcedx 52.0.0 (latest-rc)",
                "├─ salesforce-alm 53.7.2",
                "├─ data 0.4.11",
                "├─ apex 0.2.2",
                "├─ limits 1.2.1",
                "├─ schema 1.0.7",
                "├─ org 1.6.6",
                "├─ custom-metadata 1.0.12",
                "├─ user 1.3.0",
                "├─ templates 51.5.0",
                "└─ @salesforce/sfdx-plugin-lwc-test 0.1.7",
                "sfdx-cli 7.134.0 (core)",
                "source 1.8.0 (core)",
                "telemetry 1.4.0 (core)",
                "trust 1.1.0 (core)"
        ],
        "osVersion": "Linux 5.13.0-27-generic"
}
```

### Additional information

Deployment of symlinks works with version 7.130.1-fe4ce6c

```
$ sfdx update  --from-local 
Found versions: 
     7.130.1-fe4ce6c
     7.132.0-6621068
     7.134.0-4f69b20

Enter a version to update to:   7.130.1-fe4ce6c

```

**Comment by mshanemc** [comment]
please run `sfdx plugins:uninstall salesforcedx` so that you're not using super old versions of the commands.

Then try your stuff again and see if it's any better.

**Comment by vytautasgermanavicius** [comment]
@mshanemc  No, this does not help. 
I removed salesforcedx.
```
$ sfdx plugins:uninstall salesforcedx
 ›   Warning: sfdx-cli update available from 7.130.1 to 7.134.0.
Uninstalling salesforcedx... done
```
Then updated sfdx with `sfdx update`. Now i have 
```
$ sfdx version --verbose --json
{
        "cliVersion": "sfdx-cli/7.134.0",
        "architecture": "linux-x64",
        "nodeVersion": "node-v16.13.2",
        "pluginVersions": [
                "@oclif/plugin-autocomplete 0.3.0 (core)",
                "@oclif/plugin-commands 1.3.0 (core)",
                "@oclif/plugin-help 3.3.1 (core)",
                "@oclif/plugin-not-found 1.2.6 (core)",
                "@oclif/plugin-plugins 1.10.11 (core)",
                "@oclif/plugin-update 1.5.0 (core)",
                "@oclif/plugin-warn-if-update-available 1.7.3 (core)",
                "@oclif/plugin-which 1.0.4 (core)",
                "@salesforce/sfdx-plugin-lwc-test 0.1.7 (core)",
                "alias 1.2.0 (core)",
                "apex 0.7.0 (core)",
                "auth 1.8.0",
                "config 1.3.14 (core)",
                "custom-metadata 1.0.12 (core)",
                "data 0.6.8 (core)",
                "generator 1.2.1 (core)",
                "info 1.2.0 (core)",
                "limits 1.3.0 (core)",
                "org 1.11.0 (core)",
                "salesforce-alm 53.7.2 (core)",
                "schema 1.1.0 (core)",
                "sfdx-cli 7.134.0 (core)",
                "source 1.8.0 (core)",
                "telemetry 1.4.0 (core)",
                "templates 53.1.0 (core)",
                "trust 1.1.0 (core)",
                "user 1.7.0 (core)"
        ],
        "osVersion": "Linux 5.13.0-27-generic"
}
```
I still get :
```
$ sfdx force:source:deploy  -p "./vendors/SObjectFabricator"
ERROR running force:source:deploy:  /home/░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░/vendors/SObjectFabricator: Could not infer a metadata type
```

But now, reverting to 7.130.1-fe4ce6c does not help to get it working. :(

**Comment by vytautasgermanavicius** [comment]
@mshanemc  Error is in NodeFSTreeContainer.isDirectory(fsPath) . It returns false, when fsPath is symlink to directory. 
It should treat such paths as directories. 

I made following hack, i know it may be incorrect in some cases , but now I can deploy symlinked directories:
```
    isDirectory(fsPath) {
        const aaa = (0, graceful_fs_1.lstatSync)(fsPath);
        if(!aaa.isDirectory())
        {
            if(aaa.isSymbolicLink())
            {
                return this.isDirectory(graceful_fs_1.realpathSync(fsPath));
            }
            return false;
        };
        return true;
//        return (0, graceful_fs_1.lstatSync)(fsPath).isDirectory();
    }
```

**Comment by uip-robot-zz** [comment]
This issue has been linked to a new work item: [W-10494044](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000islqDYAQ/view)

**Comment by mshanemc** [comment]
@vytautasgermanavicius nice detective work there.

If you want to complete the exercise, the TS code is here: https://github.com/forcedotcom/source-deploy-retrieve/blob/aaf7cb05110a14ab9a2ef2b72d57bab77bf5859a/src/resolve/treeContainers.ts#L87

**Comment by mshanemc** [comment]
I think I found a better option

**Comment by iowillhoit** [comment]
This issue has been fixed and included in today's stable release `7.138.1`. [Update](https://developer.salesforce.com/docs/atlas.en-us.234.0.sfdx_setup.meta/sfdx_setup/sfdx_setup_update_cli.htm) to the latest `sfdx` to get these changes. Release notes can be viewed [here](https://github.com/forcedotcom/cli/tree/main/releasenotes/sfdx#71381-feb-17-2022-stable). Thank you!
